packages/vote/fpc/fpc.go

```diff
@@ -0,0 +1,267 @@
+package fpc
+
+import (
+	"container/list"
+	"context"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/iotaledger/goshimmer/packages/vote"
+	"github.com/iotaledger/hive.go/events"
+)
+
+var (
+	ErrVoteAlreadyOngoing       = errors.New("a vote is already ongoing for the given ID")
+	ErrNoOpinionGiversAvailable = errors.New("can't perform round as no opinion givers are available")
+)
+
+// New creates a new FPC instance.
+func New(opinionGiverFunc vote.OpinionGiverFunc, paras ...*Parameters) *FPC {
+	f := &FPC{
+		opinionGiverFunc: opinionGiverFunc,
+		paras:            DefaultParameters(),
+		rng:              rand.New(rand.NewSource(time.Now().UnixNano())),
+		ctxs:             make(map[string]*VoteContext),
+		queue:            list.New(),
+		queueSet:         make(map[string]struct{}),
+		events: vote.Events{
+			Finalized: events.NewEvent(vote.OpinionCaller),
+			Failed:    events.NewEvent(vote.OpinionCaller),
+			Error:     events.NewEvent(events.ErrorCaller),
+		},
+	}
+	if len(paras) > 0 {
+		f.paras = paras[0]
+	}
+	return f
+}
+
+// FPC is a DRNGRoundBasedVoter which uses the Opinion of other entities
+// in order to finalize an Opinion.
+type FPC struct {
+	events           vote.Events
+	opinionGiverFunc vote.OpinionGiverFunc
+	// the lifo queue of newly enqueued items to vote on.
+	queue *list.List
+	// contains a set of currently queued items.
+	queueSet map[string]struct{}
+	queueMu  sync.Mutex
+	// contains the set of current vote contexts.
+	ctxs   map[string]*VoteContext
+	ctxsMu sync.RWMutex
+	// parameters to use within FPC.
+	paras *Parameters
+	// indicates whether the last round was performed successfully
+	lastRoundCompletedSuccessfully bool
+	rng                            *rand.Rand
```

Comment:
Should we rename this `rng` to something like `opinionGiverRng` or `sampleRng` so that it is more clear that it's used only for randomizing the opinionGiver sample? wdyt?

Reply:
Renamed to `opinionGiverRng`